### ISSUE TITLE
Benchmarks of private code not running on MacOS

### ISSUE THREAD

**Issue by Fraser999** [nvzqz/divan#61]
I created [a small repo](https://gitlab.com/Fraser999/divan-example) to demonstrate the issue.

In the root, if I run `cargo bench` on Ubuntu, I see the expected benchmarking output, but on MacOS, I don't.

**Comment by dvdplm** [comment]
This is what I see too. The "register anywhere" feature seems to be broken on macOS (I'm on Sequoia, `15.0 (24A335)`).

**Comment by Fraser999** [comment]
> Did you by any chance dig deeper into this yet?

No, I'm afraid I've not had time.

> Do you have a hunch about where the problem lies and/or which macOS version broke it?

Also no.  For reference, I'm on Sonoma (14.6.1).  Potentially it has never worked on macOS, since the example [here](https://github.com/nvzqz/divan/blob/main/internal_benches/README.md) works for me and that's what's [used in CI](https://github.com/nvzqz/divan/blob/5478a6008666bac5f2c15cb29814c34f8c68474a/.github/workflows/ci.yml#L167-L191) to test this functionality.

**Comment by dvdplm** [comment]
Yeah the example works for me too, always did. Never figured out how to do it outside the divan crate though.

It's such a killer feature, sad it's incomplete.

**Comment by jdidion** [comment]
The best solution I've found so far is to a) create a feature flag that you can use to conditionally add `pub` to modules only during benchmarking, b) put the `divan::main()` call inside a `main` function in the `benchmarks` sub-module of whatever private module you want to benchmark, and c) call all the `main` functions from the `main` function in the `benches` folder.

So your `cargo.toml` looks like this:

```toml
[package]
name = "foo"
...

[dependencies]
divan = { version = "0.1.14", optional = true }

[features]
bench = ["dep:divan"]

[[bench]]
name = "benchmark"
harness = false
```

You have `lib.rs`:

```rust
#[cfg(not(feature = "bench"))]
mod bar;
#[cfg(not(feature = "bench"))]
mod baz;

#[cfg(feature = "bench")]
pub mod bar;
#[cfg(feature = "bench")]
pub mod baz;
```

You have `bar.rs` (`baz.rs` would be similar):

```rust
fn my_private_fn() {
   ...
}

#[cfg(feature = "bench")]
pub mod benchmarks {
    pub fn main() {
        divan::main();
    }
    
    fn bench_my_private_fn() {
        divan::black_box(super::my_private_fn());
    } 
}
```

And you have a `benches/benchmark.rs` file with the following:

```rust
extern crate foo;

fn main() {
    foo::bar::benchmarks::main();
    foo::baz::benchmarks::main();
}
```

To run benchmarks: `cargo bench --features bench`

**Comment by nvzqz** [comment]
This is quite unfortunate, but thanks @jdidion for finding a decent workaround! I'll leave this issue open until this works as expected.

**Comment by dvdplm** [comment]
@jdidion I ran into the [`visibility`](https://crates.io/crates/visibility) crate the other day, which could be helpful to make your workaround less verbose.

@nvzqz I did some debugging on this issue; I didn't get super far, but I did manage to prove to myself that the benchmark *does* make it into the `__mod_init_func` section. 
As I'm sure you are aware there has been plenty of discussion around the `#[used]` attribute and I was worried that we needed the new `#[used(linker)]` variant to prevent the compiler from optimizing out divan. This is not the case. There must be something else stopping macOS from actually calling the code before `main`, but I have not figured out why.

**Comment by nvzqz** [comment]
It appears that this issue is caused by the compiler completely discarding the crate where internal benchmarks are defined. If no items are used, then none of the crate's symbols appear in the final binary. I verified this with `otool -L`.

I attempted some approaches for internal benchmark discovery, but they didn't work:
- [`#[linkme::distributed_slice]`](https://docs.rs/linkme) produces a zero-sized slice.
- Calling `getsectiondata` on every Mach-O image header (obtained from `_dyld_image_count` + `_dyld_get_image_header`)

I'm hoping that https://github.com/rust-lang/testing-devex-team/issues/3 might fix this. cc @epage

**Comment by nvzqz** [comment]
@Fraser999 the repro code you provided now works under Rust nightly from the fix in https://github.com/rust-lang/rust/pull/133832.

However, this issue isn't entirely resolved since removing the `extern crate divan_example;` import makes it stop working.
